### DISCUSSION TITLE
Pages using searchParams do not cache

### DISCUSSION THREAD

**Discussion by joelvarty** [vercel/next.js#69943] [Help] [answered]
### Link to the code that reproduces this issue

https://github.com/joelvarty/nextjs-query-params-cache

### To Reproduce

- `yarn install`
- `yarn build`
- `yarn start`
- Browse to `http://localhost:3000/query/test`
  - notice the cache-control header specifies maxage that matches our revalidate export and stale-while-revalidate (`private, no-cache, no-store, max-age=0, must-revalidate`)
- Browse to 'http://localhost:3000/query/test'
  - notice the cache-control header specifies no-cache (`s-maxage=60, stale-while-revalidate`)

### Current vs. Expected behavior

Current:
- pages that use searchParams do NOT cache

Expected Behaviour:
- I would expect pages using searchParams to cache a different version of the page for each query param

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 18.17.1
  npm: 9.6.7
  Yarn: 1.22.19
  pnpm: N/A
Relevant Packages:
  next: 14.2.9 // Latest available version is detected (14.2.9).
  eslint-config-next: 14.2.9
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.6.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Performance

### Which stage(s) are affected? (Select all that apply)

next start (local), Vercel (Deployed), Other (Deployed)

### Additional context

This is happening in next.js version `14.2.9` currently.  I noticed it in locally and on Vercel and Netlify in a larger repo and reproduced it in a minimal repro here, which can be reproduced easily locally while checking the cache-control header..

**Comment by abhi12299** [comment]
this is not a bug: https://nextjs.org/docs/app/building-your-application/caching#dynamic-functions

**Comment by joelvarty** [comment]
Ouch.  So that means any page that needs query params can't be cached at all?

**Comment by abhi12299** [comment]
Once a page gets opted out from the full route cache, it cannot be cached by the server. It has to be rendered again.

**Comment by samcx** [comment, marked as answer]
@abhi12299 Thank you for responding!

> this is not a bug: https://nextjs.org/docs/app/building-your-application/caching#dynamic-functions

This is correct.

I asked v0 this question as well—its answer may prove helpful!

https://v0.dev/chat/nBSWSApQ0_S

**Reply by joelvarty** [reply]
Thanks!  I really appreciate that - also cool to see how v0 can answer questions like this.

**Comment by joelvarty** [comment]
I found a great workaround for this that allows you to use a specific subset of query string parameters in routes that you need to cache.

**Middleware**
- Detect the query strings (eg: `category=x, topic=y`)
- Rewrite the path with a special delimiter in the last segment that you can detect - I chose ~~~ 
- oldPath: /resources?category=x&topic=y` =>  `newPath: /resources/~~~category%3Dx%26topic%3Dy`

```
	//all the possible search params we care about...
	const paramsNames: string[] = ["subscribed", "industry", "challenge", "category", "topic", "region"]
	const otherSearchParams: { [id: string]: string; } = {}
	let hasOtherSearchParams = false

	paramsNames.forEach(paramName => {
		const val = request.nextUrl.searchParams.get(paramName)
		if (val) {
			otherSearchParams[paramName] = val
			hasOtherSearchParams = true
		}
	})

	if (hasOtherSearchParams) {
		//*** this is a request with other search params ***

		/****
			In order to preserve caching, we need to rewrite the url to a new url that includes the search params in the path.

			We will do a rewrite to the same url without the search params, except we will add a special path segement
			to the end of the path to indicate that this is a search request
			which the page can parse to get the query param values/

			eg:  /resources?category=foo&topic=bar => /resources/~~~category=%3Dfoo%26topic%3Dbar
		****/

		//get the path name
		let pathName = request.nextUrl.pathname

		//generate the special path segment
		const extraSegment = Object.keys(otherSearchParams).map(key => `${key}%3D${encodeURIComponent(otherSearchParams[key])}`).join("%26")


		//add the special path segment to a new URL
		const adjustedUrl = `${request.nextUrl.protocol}//${request.nextUrl.host}${pathName}/~~~${extraSegment}`

		console.log("PATHNAME Post: ", adjustedUrl)

		//rewrite the path
		return NextResponse.rewrite(adjustedUrl)

	}

```


**Page**
- Detect the special delimiter in the logic where you're resolving your page path
- Parse out the query string portion of the path and remove that portion from the path
- Deliver the query strings to any components that need them and use these instead of `searchParams`

```
	let globalData: any = {}

	//check if the path has a special first segment of ~~~ which means it's a got the query strings in there
	const lastSegment = params.slug.slice(-1)

	if (lastSegment[0].startsWith("~~~")) {
		const queryStr = decodeURIComponent(lastSegment[0].substring(3))

		//put the search params into the global data object so all the components can access it
		globalData.searchParams = new URLSearchParams(queryStr)

		//reset the slug to NOT have the special segment
		params.slug = params.slug.slice(0, -1)
	}
```

Here's a fairly large repo that does this (among lots of other things...)
https://github.com/agility/agility-website-next.js

**Reply by abhi12299** [reply]
If my understanding is correct, this will require a catch-all route, from which the path can be accessed and params can be extracted. Is that so?
